packages/data/src/components/with-dispatch/test/index.js

```diff
@@ -61,4 +64,31 @@ describe( 'withDispatch', () => {
 
 		expect( store.getState() ).toBe( 2 );
 	} );
+
+	it( 'return the value returned by the registry dispatch call', () => {
+		expect.assertions( 1 );
+
+		const Component = withDispatch( ( _dispatch ) => {
+			return {
+				increment: () => {
+					const actionReturnedFromDispatch = _dispatch( 'counter' ).increment();
+					expect( actionReturnedFromDispatch ).toBe( 'incremented' );
```

Comment:
Okay, now that I see what we're expecting by the `return` change, I'm not sure I like it. Why are we exposing what should be an internal implementation detail of the specific return value of the action creator?

Reply:
Because of this https://github.com/WordPress/gutenberg/pull/10089/files#diff-06dd7d804af5176903f3dd9728875d53R118 

We need the returned/saved term to be able to add it to the currently edited post. How would you solve this?

Reply:
I don't think the implementation of that component is something we should be proud of or want to encourage as a model to follow. There might be some ergonomics in being able to treat it as a promise from the component, but it quickly becomes lost with all of the ancillary requirements to account for lifecycle and unmounting. I imagine most developers would probably not even think to assign an instance property to unassign in `componentWillUnmount` and check for in the thennable of the promise.

I think the main issue is that we've implemented it such that the component is responsible for coordinating the creation and subsequent editing of post attributes, when ideally it should just be calling "add term", and all of that handled as an implementation detail of the store. The component itself should then just be a reflection of the state values for the assigned terms.

Reply:
These are two separate stores. We could potentially have a `addTermToPost` action in the `editor` store calling the more generic `addTerm` as a consequence. `addTermToPost` need to know somehow that `addTerm` is finished before continuing though, how would you propose to solve this?

Reply:
I mean, ideally it's no different than any other asynchronous store flow using controls.

Reply:
Would you be ook with adding a `registry` argument in the controls to be able to make cross-stores actions calls inside another action?

Reply:
Would a selector which composes on other stores' selectors need to refer to the registry as well, for example?

Ideally, it should, I think it's not the case at the moment (and not sure how you'd solve it as well) but yes, maybe for now calling selectors across stores is doing it wrong.

Reply:
Yes, it does seem problematic, in what I'd assumed to be identical to the problem you're outlining with needing to pass the registry to the control.

Reply:
Ok, So for now I'm not going to fix the controls (registry) ... I just updated the code to use an effect (which is the pattern we use currently in the editor). 

We should rewrite all of our effects to be actions instead (using controls) but to be able to do it, we should first add the possibility to access the registry from the controls. and this is out of scope of the current PR.

Reply:
Created this https://github.com/WordPress/gutenberg/issues/11124

Reply:
I've recently encountered some difficulty doing cross registry selectors via controls that I think is what you guys are alluding to here.  You can see my attempt (which seems to work, but I'm not sure how resilient it is yet) here: https://github.com/eventespresso/event-espresso-core/pull/637 (that particular pull has a LOT of changes so you can go to the specific data related files in the branch for the pull [here](https://github.com/eventespresso/event-espresso-core/tree/7dde346a74c0b8096fcbe30ee7b5e911f190513e/assets/src/data/eventespresso))  I may be _doing some things wrong_ in some cases, so I welcome critical feedback.  But hope it may provide practical insight into some of the challenges to overcome with the current api.